File: projects/slick/lib/slick/database/migration.rb

require "slick/registry"
require "slick/helpers"

class Slick::Database::Migration

    class << self

        include Slick::Registry

        def register(name, *args, &block)
            throw "Invalid migration name '#{name}' - it must begin with a unix timestamp" if !name.to_s.match(/\A\d+/)
            super
        end

        def schema_version
            if matches = name.to_s.match(/\A(\d+)/)
                matches[0].to_i
            else
                0
            end
        end

    end

    include Slick::Helpers

    def migrate

    end

end

File: packages/pinstripe/lib/extensions/database/migration.js

import { Class } from '../../class.js';
import { Registry } from '../../registry.js';

export const Migration = Class.extend().include({
    meta(){
        this.assignProps({ name: 'Migration' });

        this.include(Registry);

        this.assignProps({
            normalizeName: (name) => {
                if(!name.match(/^\d+/)){
                    throw new Error(`Invalid migration name '${name}' - it must begin with a unix timestamp.`);
                }
                return name;
            },

            get schemaVersion(){
                const matches = this.name.match(/^\d+/)
                if(matches){
                    return parseInt(matches[0]);
                }
                return 0;
            }
        });
    },

    initialize(database){
        this.database = database;
    }
});
